kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/KyuubiConnection.java

```diff
@@ -861,7 +861,9 @@ && hasSessionValue(AUTH_PRINCIPAL)
         && !hasSessionValue(AUTH_KYUUBI_CLIENT_KEYTAB)
         && (AUTH_KERBEROS_AUTH_TYPE_FROM_SUBJECT.equalsIgnoreCase(
                 sessConfMap.get(AUTH_KERBEROS_AUTH_TYPE))
-            || isHadoopUserGroupInformationDoAs());
+            || (!AUTH_KERBEROS_AUTH_TYPE_FROM_TICKET_CACHE.equalsIgnoreCase(
+                    sessConfMap.get(AUTH_KERBEROS_AUTH_TYPE))
+                && isHadoopUserGroupInformationDoAs()));
```

Comment:
if use specific `kerberosAuthType=fromTicketCache` we should always respect that, thus `isFromSubjectAuthMode()` should always return `false`, no matter under UGI context or not, right?

Reply:
if so, I suggest changing it to

(I use string literal in the example, please replace them with const variable)
```java
  private boolean isForciblyFromSubjectAuthMode() {
    return "fromSubject".equalsIgnoreCase(sessConfMap.get("kerberosAuthType"));
  }

  private boolean isForciblyTgtCacheAuthMode() {
    return "fromTicketCache".equalsIgnoreCase(sessConfMap.get("kerberosAuthType"));
  }

  ...

  private boolean isFromSubjectAuthMode() {
    return isSaslAuthMode()
        && hasSessionValue(AUTH_PRINCIPAL)
        && !hasSessionValue(AUTH_KYUUBI_CLIENT_PRINCIPAL)
        && !hasSessionValue(AUTH_KYUUBI_CLIENT_KEYTAB)
        && !isForciblyTgtCacheAuthMode()
        && (isForciblyFromSubjectAuthMode() || isHadoopUserGroupInformationDoAs());
  }

```

Reply:
I found another case, if the user provides `kyuubiClientPrinciple` and `kyuubiClientKeytab` while explicitly setting `kerberosAuthType=fromTicketCache`, the code path always goes keytab mode, still intuitive ...

Reply:
maybe something like
```java
  private boolean isForciblyFromKeytabAuthMode() {
    return "fromKeytab".equalsIgnoreCase(sessConfMap.get("kerberosAuthType"));
  }

  private boolean isForciblyFromSubjectAuthMode() {
    return "fromSubject".equalsIgnoreCase(sessConfMap.get("kerberosAuthType"));
  }

  private boolean isForciblyTgtCacheAuthMode() {
    return "fromTicketCache".equalsIgnoreCase(sessConfMap.get("kerberosAuthType"));
  }

  private boolean isKeytabAuthMode() {
    // handle explicit cases first
    if (isForciblyFromSubjectAuthMode() || isForciblyTgtCacheAuthMode()) {
      return false;
    }
    if (isKerberosAuthMode() && isForciblyFromKeytabAuthMode()) {
      return true;
    }
    // handle implicit cases then
    return isKerberosAuthMode()
        && hasSessionValue(AUTH_KYUUBI_CLIENT_PRINCIPAL)
        && hasSessionValue(AUTH_KYUUBI_CLIENT_KEYTAB);
  }

  private boolean isFromSubjectAuthMode() {
    // handle explicit cases first
    if (isForciblyFromKeytabAuthMode() || isForciblyTgtCacheAuthMode()) {
      return false;
    }
    if (isKerberosAuthMode() && isForciblyFromSubjectAuthMode()) {
      return true;
    }
    // handle implicit cases then
    return isKerberosAuthMode()
        && !hasSessionValue(AUTH_KYUUBI_CLIENT_KEYTAB)
        && isHadoopUserGroupInformationDoAs();
  }

  private boolean isTgtCacheAuthMode() {
    // handle explicit cases first
    if (isForciblyFromKeytabAuthMode() || isForciblyFromSubjectAuthMode()) {
      return false;
    }
    if (isKerberosAuthMode() && isForciblyTgtCacheAuthMode()) {
      return true;
    }
    // handle implicit cases then
    return isKerberosAuthMode()
        && !hasSessionValue(AUTH_KYUUBI_CLIENT_KEYTAB);
  }
```

Reply:
thanks for the suggestions, adopted.